src/Algebra/Graph/Relation/Symmetric.hs

```diff
@@ -11,27 +11,497 @@
+preSet x = R.preSet x . fromSymmetric
+
+-- | The /postset/ of an element @x@ is the set of elements that are related to
+-- it on the /right/, i.e. @postSet x == { a | xRa }@. In the context of directed
+-- graphs, this corresponds to the set of /direct successors/ of vertex @x@.
+-- Complexity: /O(n + m)/ time and /O(n)/ memory.
+--
+-- @
+-- postSet x 'empty'      == Set.'Set.empty'
+-- postSet x ('vertex' x) == Set.'Set.empty'
+-- postSet x ('edge' x y) == Set.'Set.fromList' [y]
+-- postSet 2 ('edge' 1 2) == Set.'Set.singleton' 1
+-- @
+postSet :: Ord a => a -> SymmetricRelation a -> Set.Set a
+postSet x = R.postSet x . fromSymmetric
+
+-- | The /path/ on a list of vertices.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- path []        == 'empty'
+-- path [x]       == 'vertex' x
+-- path [x,y]     == 'edges' [(x,y), (y,x)]
+-- path           == path . 'reverse'
+-- @
+path :: Ord a => [a] -> SymmetricRelation a
+path xs = case xs of []     -> empty
+                     [x]    -> vertex x
+                     (_:ys) -> edges (zip xs ys)
+
+-- | The /circuit/ on a list of vertices.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- circuit []        == 'empty'
+-- circuit [x]       == 'edge' x x
+-- circuit [x,y]     == 'edges' [(x,y), (y,x)]
+-- circuit           == circuit . 'reverse'
+-- @
+circuit :: Ord a => [a] -> SymmetricRelation a
+circuit []     = empty
+circuit (x:xs) = path $ [x] ++ xs ++ [x]
+
+-- | The /clique/ on a list of vertices.
+-- Complexity: /O((n + m) * log(n))/ time + /O(m*log(m)) time from computing the symmetricClosure and /O(n + m)/ memory.
+--
+-- @
+-- clique []         == 'empty'
+-- clique [x]        == 'vertex' x
+-- clique [x,y]      == 'edges' [(x,y), (y,x)]
+-- clique [x,y,z]    == 'edges' [(x,y), (x,z), (y,x), (y,z), (z,x), (z,y)]
+-- clique (xs ++ ys) == 'connect' (clique xs) (clique ys)
+-- clique            == clique . 'reverse'
+-- @
+clique :: Ord a => [a] -> SymmetricRelation a
+clique = SymmetricRelation . R.symmetricClosure . R.clique
+
+-- | The /biclique/ on two lists of vertices.
+-- Complexity: /O(n * log(n) + m)/ time + /O(m*log(m)) time from computing the symmetricClosure and /O(n + m)/ memory.
+--
+-- @
+-- biclique []      []      == 'empty'
+-- biclique [x]     []      == 'vertex' x
+-- biclique []      [y]     == 'vertex' y
+-- biclique [x1,x2] [y1,y2] == 'edges' [(x1,y1), (x1,y2), (y1,x1), (y1,x2), (x2,y1), (x2,y2), (y2,x1), (y2,x2)]
+-- biclique xs      ys      == 'connect' ('vertices' xs) ('vertices' ys)
+-- @
+biclique :: Ord a => [a] -> [a] -> SymmetricRelation a
+biclique xs = SymmetricRelation . R.symmetricClosure . R.biclique xs
+
+-- TODO: Optimise.
+-- | The /star/ formed by a centre vertex connected to a list of leaves.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- star x []    == 'vertex' x
+-- star x [y]   == 'edges' [(x,y), (y,x)]
```

Comment:
No need to duplicate the edge in this example. Same below.

Reply:
Already answered